app/controllers/ranks_controller.rb

```diff
@@ -0,0 +1,17 @@
+
+class RanksController < ApplicationController
+  def update_ranks
+
+    params['users'].each_pair do |identifier, rank_data|
+      identity = Identity.find_by(identifier: identifier)
+      next unless identity
+
+      user = identity.user
+      next unless user
+
+      Rank.where(user_id: user.id).delete_all
+      Rank.create(name: rank_data['group'], user_id: user.id, realm: params['realm'])
+    end
+    render plain: 'Ranks updated', status: 202
+  end
```

Comment:
This would probably end up being a _ton_ of queries run sequentially.

Because we're on rails 5 still, we'll need to use a gem that will let us bulk update/insert ("upsert")

This is a good one: https://github.com/zdennis/activerecord-import/

And we can use this feature to make our lives easy: https://github.com/zdennis/activerecord-import/#duplicate-key-update

It'll let us update rank if they already exist for this realm, or create new ones if they don't - all in one-ish query!

Reply:
so should realm + rank have a unique constraint?
would mean you can only have one rank per realm.

Reply:
Yep, good idea 👍

Reply:
we will still need at least 1 select query for every user

Reply:
I reckon you can batch that, too.

Collect a list of identifiers, do one query with that, and then work out which ones we need to upsert based on the first query.

I think this entire thing can be done with <3 queries

Reply:
```rb
identities = params['users'].keys
existing_identities = Identity.where(identifier: identities).pluck(:identifier)

upserts = []
existing_identities.each do |identifier|

Reply:
Something like this, maybe

```rb
users = params['users']
realm = params['realm']

identities = users.keys
existing_identities = Identity.includes(:user).where(identifier: identities)

upserts = []
existing_identities.each do |identity|
  identifier = identity.identifier
  user = identity.user

  rank_data = users[identifier]
  upserts.push(user.ranks.new(name: rank_data['group'], realm: realm))
end

Rank.import upserts, on_duplicate_key_update: [:name]
```